src/reference/mod.rs

```diff
@@ -0,0 +1,404 @@
+//! syntax node.)
+//!
+//! ``` tsg
+//! (identifier) @id
+//! {
+//!   let local_variable = "a string"
+//!   ; The following would be an error, since `let` variables are immutable:
+//!   ; set local_variable = "a new value"
+//!
+//!   ; The following is also an error, since you can't mutate a variable that
+//!   ; doesn't exist:
+//!   ; set missing_variable = 42
+//!
+//!   var mutable_variable = "first value"
+//!   set mutable_variable = "second value"
+//!
+//!   var @id::kind = "id"
+//! }
+//! ```
+//!
+//! Variables can be referenced anywhere that you can provide an expression.  It's an error if you
+//! try to reference a variable that hasn't been defined yet.  (Remember that stanzas are processed
+//! in the order they appear in the file, and each stanza's matches are processed in the order they
+//! appear in the syntax tree.)
+//!
+//! # Functions
+//!
+//! The process executing a graph DSL file can provide **_functions_** that can be called from
+//! within graph DSL stanzas.
+//!
+//! Function calls use a Lisp-like syntax, where the name of the function being called is _inside_
+//! of the parentheses.  The parameters to a function call are arbitrary expressions.  For
+//! instance, if the executing process provides a function named `+`, you could call it as:
+//!
+//! ``` tsg
+//! (identifier) @id
+//! {
+//!    let x = 4
+//!    let @id::nine = (+ x 5)
+//! }
+//! ```
+//!
+//! Note that it's the process executing the graph DSL file that decides which functions are
+//! available.  We do define a [standard library][], and most of the time those are the functions
+//! that are available, but you should double-check the documentation of whatever graph DSL tool
+//! you're using to make sure.
+//!
+//! [standard library]: functions/index.html
+//!
+//! # Graph nodes
+//!
+//! You can use this graph DSL to create any graph structure that you want.  There are no
+//! limitations on which graph nodes you create, nor on how you use edges to connect the graph
+//! nodes.  You are not limited to creating a tree, and in particular, you are not limited to
+//! creating a tree that "lines" up with the parsed syntax tree.
+//!
+//! You use the [`node`][] function to create a new graph node.  The function returns a reference
+//! to the new node.  You will almost always assign that result to a variable so that you can then
+//! do interesting things with the node:
+//!
+//! [`node`]: functions/index.html#node
+//!
+//! ``` tsg
+//! (identifier) @id
+//! {
+//!   let new_node = (node)
+//! }
+//! ```
+//!
+//! Graph nodes are usually attached to syntax nodes using [scoped variables](#variables), so that
+//! you can refer to them from multiple stanzas:
+//!
+//! ``` tsg
+//! (identifier) @id
+//! {
+//!   let @id::node = (node)
+//! }
+//!
+//! (dotted_name (identifier) @dotted_element)
+//! {
+//!   ; We will learn more about the attr statement below
+//!   attr (@dotted_element::node) kind = "dotted"
```

Comment:
Clarifying question: are the parenthesis around `(@dotted_element::node)` required?

Reply:
Yes, as currently written, but for aesthetic reasons not to make the parsing easier.

I had put that in more because of the edge attribute statement, so that it's clearer which part of the statement is the edge reference and which part is the list of attributes:

```
attr (@id::source -> @id::sink) precedence = 10
```

and then thought I should keep the parens for node attributes, too, for consistency.